pkg/metrics/metrics.go

```diff
@@ -0,0 +1,209 @@
+	awsThrottleMetric = "cloudprovider_aws_api_throttled_requests_total"
+	awsErrorMetric    = "cloudprovider_aws_api_request_errors"
+
+	// Internal driver metrics
+	internalDurationMetric = "ebs_csi_aws_com_duration_seconds"
+	internalValueMetric    = "ebs_csi_aws_com_operation"
+	internalErrorMetric    = "ebs_csi_aws_com_errors_total"
+)
+
+var (
+	// r is the singleton instance of metricRecorder.
+	r *metricRecorder
+
+	// initialize ensures that metricRecorder is initialized only once.
+	initialize sync.Once
+)
+
+// metricType is an enum for the different types of metrics.
+type metricType int
+
+// metricRecorder provides an organized way of recording and managing metrics related to both AWS and driver operations.
+type metricRecorder struct {
+	// registry is the central store where all metrics are registered and exposed for collection.
+	registry metrics.KubeRegistry
+
+	// metrics maps a metric type (AWS or Internal) to its associated measurement data -- duration, values, error counts, and throttling events.
+	metrics map[metricType]metricFields
+}
+
+// metricFields holds the metrics for a specific metric type.
+type metricFields struct {
+	duration *metrics.HistogramVec
+	value    *metrics.HistogramVec
+	err      *metrics.CounterVec
+	throttle *metrics.CounterVec
+}
+
+// Recorder returns the singleton instance of metricRecorder.
+// Metrics are registered only once throughout the driver's lifetime.
+func Recorder() *metricRecorder {
+	initialize.Do(func() {
+		r = newRecorder()
+		r.registerMetrics()
+	})
+	return r
+}
+
+// Duration logs the duration of an operation for the specified metric type.
+//
+// Parameters:
+// - t: The type of the metric (e.g., AWS or Internal).
+// - o: The name of the operation for which the metric is being recorded.
+// - d: The duration taken for the action in seconds.
+func (m *metricRecorder) Duration(t metricType, o string, d float64) {
+	if metric, ok := m.metrics[t]; ok {
+		metric.duration.With(metrics.Labels{"request": o}).Observe(d)
+	} else {
+		klog.Warningf("Metric type %d not found", t)
+	}
+}
+
+// Error logs an error for the specified metric type.
+//
+// Parameters:
+// - t: The type of the metric.
+// - o: The name of the operation where the error occurred.
+func (m *metricRecorder) Error(t metricType, o string) {
+	if metric, ok := m.metrics[t]; ok {
+		metric.err.With(metrics.Labels{"request": o}).Inc()
+	} else {
+		klog.Warningf("Metric type %d not found", t)
+	}
+}
+
+// Throttle logs a throttling event for the metric type.
+//
+// Parameters:
+// - t: The type of the metric.
+// - o: The name of the operation that was throttled.
+func (m *metricRecorder) Throttle(t metricType, o string) {
+	if metric, ok := m.metrics[t]; ok {
+		metric.throttle.With(metrics.Labels{"operation_name": o}).Inc()
+	} else {
+		klog.Warningf("Metric type %d not found", t)
+	}
+}
+
+// Value logs a value observation for the specified metric type.
+// Parameters:
+// - t: The type of the metric.
```

Comment:
major np (doesn't require a new PR): The Go-way of writing docstrings is to put named arguments directly in the comment. See: https://tip.golang.org/doc/comment